// ngx-feedback/esm5/lib/services/component-loader.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, ComponentFactoryResolver, ApplicationRef, Injector } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * @template T
 */
var ComponentLoaderService = /** @class */ (function () {
    function ComponentLoaderService(componentFactoryResolver, appRef, injector) {
        this.componentFactoryResolver = componentFactoryResolver;
        this.appRef = appRef;
        this.injector = injector;
    }
    /**
     * @param {?} componentType
     * @return {?}
     */
    ComponentLoaderService.prototype.load = /**
     * @param {?} componentType
     * @return {?}
     */
    function (componentType) {
        // 1. Create a component reference from the component
        /** @type {?} */
        var ref = this.componentFactoryResolver.resolveComponentFactory(componentType).create(this.injector);
        // 2. Attach component to the appRef so that it's inside the ng component tree
        this.appRef.attachView(ref.hostView);
        // 3. Get DOM element from component
        /** @type {?} */
        var domElem = (/** @type {?} */ (((/** @type {?} */ (ref.hostView))).rootNodes[0]));
        // 4. Append DOM element to the body
        document.body.appendChild(domElem);
        ref.changeDetectorRef.detectChanges();
        return ref;
    };
    ComponentLoaderService.decorators = [
        { type: Injectable, args: [{ providedIn: 'root' },] }
    ];
    /** @nocollapse */
    ComponentLoaderService.ctorParameters = function () { return [
        { type: ComponentFactoryResolver },
        { type: ApplicationRef },
        { type: Injector }
    ]; };
    /** @nocollapse */ ComponentLoaderService.ngInjectableDef = i0.defineInjectable({ factory: function ComponentLoaderService_Factory() { return new ComponentLoaderService(i0.inject(i0.ComponentFactoryResolver), i0.inject(i0.ApplicationRef), i0.inject(i0.INJECTOR)); }, token: ComponentLoaderService, providedIn: "root" });
    return ComponentLoaderService;
}());
export { ComponentLoaderService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ComponentLoaderService.prototype.componentFactoryResolver;
    /**
     * @type {?}
     * @private
     */
    ComponentLoaderService.prototype.appRef;
    /**
     * @type {?}
     * @private
     */
    ComponentLoaderService.prototype.injector;
}
