// ngx-feedback/esm5/lib/services/events.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
var EventsService = /** @class */ (function () {
    function EventsService() {
        this.feedbackClick = new Subject();
        this.feedbackClickObservable = this.feedbackClick.asObservable();
        this.specificFeedbackClick = new Subject();
        this.specificFeedbackClickObservable = this.specificFeedbackClick.asObservable();
        this.genericFeedbackClick = new Subject();
        this.genericFeedbackClickObservable = this.genericFeedbackClick.asObservable();
        this.closeClick = new Subject();
        this.closeClickClickObservable = this.closeClick.asObservable();
        this.onSendClick = new Subject();
        this.onSendClickObservable = this.onSendClick.asObservable();
    }
    /**
     * @return {?}
     */
    EventsService.prototype.onSpecificFeedbackClick = /**
     * @return {?}
     */
    function () {
        this.specificFeedbackClick.next();
    };
    /**
     * @return {?}
     */
    EventsService.prototype.onGenericFeedbackClik = /**
     * @return {?}
     */
    function () {
        this.genericFeedbackClick.next();
    };
    /**
     * @return {?}
     */
    EventsService.prototype.onFeedbackButtonClick = /**
     * @return {?}
     */
    function () {
        this.feedbackClick.next();
    };
    /**
     * @return {?}
     */
    EventsService.prototype.onCloseClicked = /**
     * @return {?}
     */
    function () {
        this.closeClick.next();
    };
    /**
     * @param {?} data
     * @return {?}
     */
    EventsService.prototype.onSendClicked = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        this.onSendClick.next(data);
    };
    EventsService.decorators = [
        { type: Injectable, args: [{ providedIn: 'root' },] }
    ];
    /** @nocollapse */
    EventsService.ctorParameters = function () { return []; };
    /** @nocollapse */ EventsService.ngInjectableDef = i0.defineInjectable({ factory: function EventsService_Factory() { return new EventsService(); }, token: EventsService, providedIn: "root" });
    return EventsService;
}());
export { EventsService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    EventsService.prototype.feedbackClick;
    /** @type {?} */
    EventsService.prototype.feedbackClickObservable;
    /**
     * @type {?}
     * @private
     */
    EventsService.prototype.specificFeedbackClick;
    /** @type {?} */
    EventsService.prototype.specificFeedbackClickObservable;
    /**
     * @type {?}
     * @private
     */
    EventsService.prototype.genericFeedbackClick;
    /** @type {?} */
    EventsService.prototype.genericFeedbackClickObservable;
    /**
     * @type {?}
     * @private
     */
    EventsService.prototype.closeClick;
    /** @type {?} */
    EventsService.prototype.closeClickClickObservable;
    /**
     * @type {?}
     * @private
     */
    EventsService.prototype.onSendClick;
    /** @type {?} */
    EventsService.prototype.onSendClickObservable;
}
