// ngx-feedback/esm5/lib/services/modal.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { ComponentLoaderService } from './component-loader.service';
import { ModalComponent } from '../components/modal/modal.component';
import { BackdropComponent } from '../components/backdrop/backdrop.component';
import * as i0 from "@angular/core";
import * as i1 from "./component-loader.service";
var ModalService = /** @class */ (function () {
    function ModalService(componentLoader) {
        this.componentLoader = componentLoader;
    }
    /**
     * @param {?} preview
     * @return {?}
     */
    ModalService.prototype.openWithPreview = /**
     * @param {?} preview
     * @return {?}
     */
    function (preview) {
        this.draw();
        this.modalRef.instance.setPreview(preview);
    };
    /**
     * @return {?}
     */
    ModalService.prototype.draw = /**
     * @return {?}
     */
    function () {
        this.backdropRef = this.componentLoader.load(BackdropComponent);
        this.modalRef = this.componentLoader.load(ModalComponent);
    };
    /**
     * @return {?}
     */
    ModalService.prototype.remove = /**
     * @return {?}
     */
    function () {
        this.modalRef.destroy();
        this.backdropRef.destroy();
    };
    ModalService.decorators = [
        { type: Injectable, args: [{ providedIn: 'root' },] }
    ];
    /** @nocollapse */
    ModalService.ctorParameters = function () { return [
        { type: ComponentLoaderService }
    ]; };
    /** @nocollapse */ ModalService.ngInjectableDef = i0.defineInjectable({ factory: function ModalService_Factory() { return new ModalService(i0.inject(i1.ComponentLoaderService)); }, token: ModalService, providedIn: "root" });
    return ModalService;
}());
export { ModalService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ModalService.prototype.modalRef;
    /**
     * @type {?}
     * @private
     */
    ModalService.prototype.backdropRef;
    /**
     * @type {?}
     * @private
     */
    ModalService.prototype.componentLoader;
}
