// ngx-feedback/esm5/lib/services/orchestrator.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { ModalService } from './modal.service';
import { SelectorsService } from './selectors.service';
import { EventsService } from './events.service';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "./modal.service";
import * as i2 from "./selectors.service";
import * as i3 from "./events.service";
var OrchestratorService = /** @class */ (function () {
    function OrchestratorService(modalService, selectorsService, eventsService) {
        this.modalService = modalService;
        this.selectorsService = selectorsService;
        this.eventsService = eventsService;
        this.subscriptions = [];
        this.onSend = new Subject();
    }
    /**
     * @return {?}
     */
    OrchestratorService.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.subscriptions.push(this.onFeedbackButtonClick(), this.onSpecificFeedbackClick(), this.onGenericFeedbackClick(), this.onElementSelected(), this.onCloseClick(), this.onSendClick());
    };
    /**
     * @private
     * @return {?}
     */
    OrchestratorService.prototype.onElementSelected = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        return this.selectorsService.preview.subscribe((/**
         * @param {?} prev
         * @return {?}
         */
        function (prev) {
            _this.modalService.openWithPreview(prev);
        }));
    };
    /**
     * @private
     * @return {?}
     */
    OrchestratorService.prototype.onFeedbackButtonClick = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        return this.eventsService.feedbackClickObservable.subscribe((/**
         * @return {?}
         */
        function () {
            _this.modalService.draw();
        }));
    };
    /**
     * @private
     * @return {?}
     */
    OrchestratorService.prototype.onSpecificFeedbackClick = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        return this.eventsService.specificFeedbackClickObservable.subscribe((/**
         * @return {?}
         */
        function () {
            _this.modalService.remove();
            _this.selectorsService.elementsHighlight();
        }));
    };
    /**
     * @private
     * @return {?}
     */
    OrchestratorService.prototype.onGenericFeedbackClick = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        return this.eventsService.genericFeedbackClickObservable.subscribe((/**
         * @return {?}
         */
        function () {
            _this.modalService.remove();
            _this.selectorsService.pageScreenshot();
        }));
    };
    /**
     * @private
     * @return {?}
     */
    OrchestratorService.prototype.onCloseClick = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        return this.eventsService.closeClickClickObservable.subscribe((/**
         * @return {?}
         */
        function () {
            _this.modalService.remove();
        }));
    };
    /**
     * @private
     * @return {?}
     */
    OrchestratorService.prototype.onSendClick = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        return this.eventsService.onSendClickObservable.subscribe((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            _this.onSend.next(data);
        }));
    };
    /**
     * @return {?}
     */
    OrchestratorService.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.subscriptions.forEach((/**
         * @param {?} sub
         * @return {?}
         */
        function (sub) {
            sub.unsubscribe();
        }));
    };
    OrchestratorService.decorators = [
        { type: Injectable, args: [{ providedIn: 'root' },] }
    ];
    /** @nocollapse */
    OrchestratorService.ctorParameters = function () { return [
        { type: ModalService },
        { type: SelectorsService },
        { type: EventsService }
    ]; };
    /** @nocollapse */ OrchestratorService.ngInjectableDef = i0.defineInjectable({ factory: function OrchestratorService_Factory() { return new OrchestratorService(i0.inject(i1.ModalService), i0.inject(i2.SelectorsService), i0.inject(i3.EventsService)); }, token: OrchestratorService, providedIn: "root" });
    return OrchestratorService;
}());
export { OrchestratorService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    OrchestratorService.prototype.subscriptions;
    /** @type {?} */
    OrchestratorService.prototype.onSend;
    /**
     * @type {?}
     * @private
     */
    OrchestratorService.prototype.modalService;
    /**
     * @type {?}
     * @private
     */
    OrchestratorService.prototype.selectorsService;
    /**
     * @type {?}
     * @private
     */
    OrchestratorService.prototype.eventsService;
}
