// ngx-feedback/esm5/lib/services/selectors.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, RendererFactory2 } from '@angular/core';
import { fromEvent, Subject } from 'rxjs';
import html2canvas from 'html2canvas';
import * as i0 from "@angular/core";
var SelectorsService = /** @class */ (function () {
    function SelectorsService(rendererFactory) {
        this.rendererFactory = rendererFactory;
        this.preview = new Subject();
        this.subs = [];
        this.renderer = this.rendererFactory.createRenderer(null, null);
    }
    /**
     * @return {?}
     */
    SelectorsService.prototype.pageScreenshot = /**
     * @return {?}
     */
    function () {
        this.screenshot(document.body);
    };
    /**
     * @return {?}
     */
    SelectorsService.prototype.elementsHighlight = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var element;
        /** @type {?} */
        var elements = [];
        this.source = fromEvent(document, 'mousemove').subscribe((/**
         * @param {?} ev
         * @return {?}
         */
        function (ev) {
            if (element) {
                _this.renderer.removeClass(element, 'ngx-feedback__highlight');
            }
            element = document.elementFromPoint(ev.x, ev.y);
            if (elements.findIndex((/**
             * @param {?} el
             * @return {?}
             */
            function (el) { return el === element; })) === -1) {
                elements.push(element);
                _this.subs.push(fromEvent(element, 'click').subscribe((/**
                 * @param {?} evt
                 * @return {?}
                 */
                function (evt) {
                    evt.stopPropagation();
                    _this.screenshot(element);
                })));
            }
            _this.renderer.addClass(element, 'ngx-feedback__highlight');
        }));
    };
    /**
     * @param {?} element
     * @return {?}
     */
    SelectorsService.prototype.screenshot = /**
     * @param {?} element
     * @return {?}
     */
    function (element) {
        var _this = this;
        /** @type {?} */
        var image;
        html2canvas(element).then((/**
         * @param {?} canvas
         * @return {?}
         */
        function (canvas) {
            image = ((/** @type {?} */ (canvas))).toDataURL();
            _this.preview.next(image);
        }));
        this.renderer.removeClass(element, 'ngx-feedback__highlight');
        this.clean();
    };
    /**
     * @return {?}
     */
    SelectorsService.prototype.clean = /**
     * @return {?}
     */
    function () {
        if (this.source) {
            this.source.unsubscribe();
        }
        this.subs.forEach((/**
         * @param {?} sub
         * @return {?}
         */
        function (sub) { return sub.unsubscribe(); }));
    };
    SelectorsService.decorators = [
        { type: Injectable, args: [{ providedIn: 'root' },] }
    ];
    /** @nocollapse */
    SelectorsService.ctorParameters = function () { return [
        { type: RendererFactory2 }
    ]; };
    /** @nocollapse */ SelectorsService.ngInjectableDef = i0.defineInjectable({ factory: function SelectorsService_Factory() { return new SelectorsService(i0.inject(i0.RendererFactory2)); }, token: SelectorsService, providedIn: "root" });
    return SelectorsService;
}());
export { SelectorsService };
if (false) {
    /** @type {?} */
    SelectorsService.prototype.preview;
    /** @type {?} */
    SelectorsService.prototype.subs;
    /** @type {?} */
    SelectorsService.prototype.source;
    /**
     * @type {?}
     * @private
     */
    SelectorsService.prototype.renderer;
    /**
     * @type {?}
     * @private
     */
    SelectorsService.prototype.rendererFactory;
}
