// ngx-feedback/esm5/lib/components/backdrop/backdrop.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Output } from '@angular/core';
var BackdropComponent = /** @class */ (function () {
    function BackdropComponent() {
        this.click = new EventEmitter();
    }
    /**
     * @return {?}
     */
    BackdropComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @return {?}
     */
    BackdropComponent.prototype.onClick = /**
     * @return {?}
     */
    function () {
        this.click.emit();
    };
    BackdropComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ngx-feedback-backdrop',
                    template: "<div class=\"ngx-feedback-backdrop__container\"></div>",
                    styles: [".ngx-feedback-backdrop__container{position:fixed;top:0;left:0;width:100%;height:100%;z-index:1000;background-color:#3b5560;opacity:.7}"]
                }] }
    ];
    /** @nocollapse */
    BackdropComponent.ctorParameters = function () { return []; };
    BackdropComponent.propDecorators = {
        click: [{ type: Output }]
    };
    return BackdropComponent;
}());
export { BackdropComponent };
if (false) {
    /** @type {?} */
    BackdropComponent.prototype.click;
}
