// ngx-feedback/esm5/lib/components/star-vote/star-vote.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Output, EventEmitter } from '@angular/core';
var StarVoteComponent = /** @class */ (function () {
    function StarVoteComponent() {
        this.stars = [
            { type: 'empty', value: 1 },
            { type: 'empty', value: 2 },
            { type: 'empty', value: 3 },
            { type: 'empty', value: 4 },
            { type: 'empty', value: 5 }
        ];
        this.voted = new EventEmitter();
    }
    /**
     * @return {?}
     */
    StarVoteComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @param {?} index
     * @return {?}
     */
    StarVoteComponent.prototype.onClick = /**
     * @param {?} index
     * @return {?}
     */
    function (index) {
        this.stars.forEach((/**
         * @param {?} s
         * @param {?} i
         * @return {?}
         */
        function (s, i) {
            if (i <= index) {
                s.type = 'full';
            }
            else {
                s.type = 'empty';
            }
        }));
        this.voted.emit(index + 1);
    };
    /**
     * @param {?} index
     * @return {?}
     */
    StarVoteComponent.prototype.onMouseEnter = /**
     * @param {?} index
     * @return {?}
     */
    function (index) {
        this.stars.forEach((/**
         * @param {?} s
         * @param {?} i
         * @return {?}
         */
        function (s, i) {
            if (i <= index && s.type !== 'full') {
                s.type = 'temp';
            }
        }));
    };
    /**
     * @param {?} index
     * @return {?}
     */
    StarVoteComponent.prototype.onMouseLeave = /**
     * @param {?} index
     * @return {?}
     */
    function (index) {
        this.stars.forEach((/**
         * @param {?} s
         * @param {?} i
         * @return {?}
         */
        function (s, i) {
            if (s.type === 'temp') {
                s.type = 'empty';
            }
        }));
    };
    StarVoteComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ngx-feedback-star-vote',
                    template: "<div class=\"ngx-feedback-star-vote__container\">\n    <div class=\"ngx-feedback-star-vote__star\" *ngFor=\"let s of stars; index as i\">\n        <ngx-feedback-star\n            (click)=\"onClick(i)\"\n            (mouseenter)=\"onMouseEnter(i)\"\n            (mouseleave)=\"onMouseLeave(i)\"\n            [type]=\"s.type\"\n        ></ngx-feedback-star>\n    </div>\n</div>\n",
                    styles: [".ngx-feedback-star-vote__container{display:flex;flex-direction:row;justify-content:center;padding:10px}"]
                }] }
    ];
    /** @nocollapse */
    StarVoteComponent.ctorParameters = function () { return []; };
    StarVoteComponent.propDecorators = {
        voted: [{ type: Output }]
    };
    return StarVoteComponent;
}());
export { StarVoteComponent };
if (false) {
    /** @type {?} */
    StarVoteComponent.prototype.stars;
    /** @type {?} */
    StarVoteComponent.prototype.voted;
}
