// ngx-feedback/esm5/lib/ngx-feedback.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { ComponentLoaderService } from './services/component-loader.service';
import { FeedbackButtonComponent } from './components/feedback-button/feedback-button.component';
import { OrchestratorService } from './services';
import * as i0 from "@angular/core";
import * as i1 from "./services/component-loader.service";
import * as i2 from "./services/orchestrator.service";
var NgxFeedbackService = /** @class */ (function () {
    function NgxFeedbackService(componentLoader, orchestratorService) {
        this.componentLoader = componentLoader;
        this.orchestratorService = orchestratorService;
    }
    /**
     * @return {?}
     */
    NgxFeedbackService.prototype.listenForFeedbacks = /**
     * @return {?}
     */
    function () {
        this.orchestratorService.ngOnInit();
        this.componentLoader.load(FeedbackButtonComponent);
        return this.orchestratorService.onSend;
    };
    NgxFeedbackService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    NgxFeedbackService.ctorParameters = function () { return [
        { type: ComponentLoaderService },
        { type: OrchestratorService }
    ]; };
    /** @nocollapse */ NgxFeedbackService.ngInjectableDef = i0.defineInjectable({ factory: function NgxFeedbackService_Factory() { return new NgxFeedbackService(i0.inject(i1.ComponentLoaderService), i0.inject(i2.OrchestratorService)); }, token: NgxFeedbackService, providedIn: "root" });
    return NgxFeedbackService;
}());
export { NgxFeedbackService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    NgxFeedbackService.prototype.componentLoader;
    /**
     * @type {?}
     * @private
     */
    NgxFeedbackService.prototype.orchestratorService;
}
