// ngx-feedback/esm5/lib/components/feedback-button/feedback-button.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewEncapsulation } from '@angular/core';
import { EventsService } from '../../services/events.service';
var FeedbackButtonComponent = /** @class */ (function () {
    function FeedbackButtonComponent(eventsService) {
        this.eventsService = eventsService;
    }
    /**
     * @return {?}
     */
    FeedbackButtonComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @return {?}
     */
    FeedbackButtonComponent.prototype.onClick = /**
     * @return {?}
     */
    function () {
        this.eventsService.onFeedbackButtonClick();
    };
    FeedbackButtonComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ngx-feedback-main-btn',
                    template: "<button (click)=\"onClick()\" class=\"feedback-button__container\">Feedback</button>\n",
                    encapsulation: ViewEncapsulation.None,
                    styles: [".feedback-button__container{position:fixed;right:-34px;top:calc(50% - 37.5px);cursor:pointer;-webkit-transform:rotateZ(90deg);transform:rotateZ(-90deg);border:none;border-radius:5px 5px 0 0;background-color:#3b5560;color:#fff;font-size:14px;padding:5px 10px;font-weight:700;font-family:Verdana;outline:0}.ngx-feedback__highlight{outline:#ffc058 solid 4px;pointer-events:none;cursor:default!important}"]
                }] }
    ];
    /** @nocollapse */
    FeedbackButtonComponent.ctorParameters = function () { return [
        { type: EventsService }
    ]; };
    return FeedbackButtonComponent;
}());
export { FeedbackButtonComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    FeedbackButtonComponent.prototype.eventsService;
}
