// ngx-feedback/esm2015/ngx-feedback.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */
export { NgxFeedbackService, NgxFeedbackModule, FeedbackData } from './public-api';
export { BackdropComponent as ɵg } from './lib/components/backdrop/backdrop.component';
export { ButtonComponent as ɵf } from './lib/components/button/button.component';
export { FeedbackButtonComponent as ɵc } from './lib/components/feedback-button/feedback-button.component';
export { ModalComponent as ɵe } from './lib/components/modal/modal.component';
export { StarVoteComponent as ɵi } from './lib/components/star-vote/star-vote.component';
export { StarComponent as ɵh } from './lib/components/star/star.component';
export { OrchestratorService as ɵb } from './lib/services';
export { ComponentLoaderService as ɵa } from './lib/services/component-loader.service';
export { EventsService as ɵd } from './lib/services/events.service';
export { ModalService as ɵk } from './lib/services/modal.service';
export { OrchestratorService as ɵj } from './lib/services/orchestrator.service';
export { SelectorsService as ɵl } from './lib/services/selectors.service';
