// ngx-feedback/esm5/lib/components/star/star.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
var StarComponent = /** @class */ (function () {
    function StarComponent() {
        this.click = new EventEmitter();
    }
    /**
     * @return {?}
     */
    StarComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @return {?}
     */
    StarComponent.prototype.onClick = /**
     * @return {?}
     */
    function () {
        this.click.emit();
    };
    StarComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ngx-feedback-star',
                    template: "<div class=\"ngx-feedback-star__container\" (click)=\"onClick()\">\n    <div *ngIf=\"type === 'full' || type === 'temp'; else empty\">\n        <svg xmlns=\"http://www.w3.org/2000/svg\" width=\"40\" height=\"40\" viewBox=\"0 0 24 24\" fill=\"#ffc058\">\n            <path\n                d=\"M12 .587l3.668 7.568 8.332 1.151-6.064 5.828 1.48 8.279-7.416-3.967-7.417 3.967 1.481-8.279-6.064-5.828 8.332-1.151z\"\n            />\n        </svg>\n    </div>\n    <ng-template #empty>\n        <div>\n            <svg xmlns=\"http://www.w3.org/2000/svg\" width=\"40\" height=\"40\" viewBox=\"0 0 24 24\" fill=\"#ffc058\">\n                <path\n                    d=\"M12 5.173l2.335 4.817 5.305.732-3.861 3.71.942 5.27-4.721-2.524-4.721 2.525.942-5.27-3.861-3.71 5.305-.733 2.335-4.817zm0-4.586l-3.668 7.568-8.332 1.151 6.064 5.828-1.48 8.279 7.416-3.967 7.416 3.966-1.48-8.279 6.064-5.827-8.332-1.15-3.668-7.569z\"\n                />\n            </svg>\n        </div>\n    </ng-template>\n</div>\n",
                    styles: [".ngx-feedback-star__container{cursor:pointer;padding:5px}"]
                }] }
    ];
    /** @nocollapse */
    StarComponent.ctorParameters = function () { return []; };
    StarComponent.propDecorators = {
        type: [{ type: Input }],
        click: [{ type: Output }]
    };
    return StarComponent;
}());
export { StarComponent };
if (false) {
    /** @type {?} */
    StarComponent.prototype.type;
    /** @type {?} */
    StarComponent.prototype.click;
}
