// @ngu/carousel/esm2022/lib/ngu-carousel/ngu-carousel-hammer-manager.mjs
import { Injectable } from '@angular/core';
import { Observable, Subject, defer, fromEvent, map, shareReplay, takeUntil } from 'rxjs';
import * as i0 from "@angular/core";
export class NguHammerLoader {
    constructor() {
        this._hammer$ = defer(() => import('hammerjs')).pipe(shareReplay({ bufferSize: 1, refCount: true }));
    }
    load() {
        return this._hammer$;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.4", ngImport: i0, type: NguHammerLoader, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "18.0.4", ngImport: i0, type: NguHammerLoader, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.4", ngImport: i0, type: NguHammerLoader, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }] });
export class NguCarouselHammerManager {
    constructor(_ngZone, _nguHammerLoader) {
        this._ngZone = _ngZone;
        this._nguHammerLoader = _nguHammerLoader;
        this._destroy$ = new Subject();
    }
    ngOnDestroy() {
        this._destroy$.next();
    }
    createHammer(element) {
        return this._nguHammerLoader.load().pipe(map(() => 
        // Note: The Hammer manager should be created outside of the Angular zone since it sets up
        //       `pointermove` event listener which triggers change detection every time the pointer is moved.
        this._ngZone.runOutsideAngular(() => new Hammer(element))), 
        // Note: the dynamic import is always a microtask which may run after the view is destroyed.
        //       `takeUntil` is used to prevent setting Hammer up if the view had been destroyed before
        //       the HammerJS is loaded.
        takeUntil(this._destroy$));
    }
    on(hammer, event) {
        return fromEvent(hammer, event).pipe(
        // Note: We have to re-enter the Angular zone because Hammer would trigger events outside of the
        //       Angular zone (since we set it up with `runOutsideAngular`).
        enterNgZone(this._ngZone), takeUntil(this._destroy$));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.4", ngImport: i0, type: NguCarouselHammerManager, deps: [{ token: i0.NgZone }, { token: NguHammerLoader }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "18.0.4", ngImport: i0, type: NguCarouselHammerManager }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.4", ngImport: i0, type: NguCarouselHammerManager, decorators: [{
            type: Injectable
        }], ctorParameters: () => [{ type: i0.NgZone }, { type: NguHammerLoader }] });
function enterNgZone(ngZone) {
    return (source) => new Observable(subscriber => source.subscribe({
        next: value => ngZone.run(() => subscriber.next(value)),
        error: error => ngZone.run(() => subscriber.error(error)),
        complete: () => ngZone.run(() => subscriber.complete())
    }));
}
