// ngx-feedback/esm2015/lib/components/backdrop/backdrop.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Output } from '@angular/core';
export class BackdropComponent {
    constructor() {
        this.click = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() { }
    /**
     * @return {?}
     */
    onClick() {
        this.click.emit();
    }
}
BackdropComponent.decorators = [
    { type: Component, args: [{
                selector: 'ngx-feedback-backdrop',
                template: "<div class=\"ngx-feedback-backdrop__container\"></div>",
                styles: [".ngx-feedback-backdrop__container{position:fixed;top:0;left:0;width:100%;height:100%;z-index:1000;background-color:#3b5560;opacity:.7}"]
            }] }
];
/** @nocollapse */
BackdropComponent.ctorParameters = () => [];
BackdropComponent.propDecorators = {
    click: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    BackdropComponent.prototype.click;
}
