// ngx-feedback/esm2015/lib/services/modal.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { ComponentLoaderService } from './component-loader.service';
import { ModalComponent } from '../components/modal/modal.component';
import { BackdropComponent } from '../components/backdrop/backdrop.component';
import * as i0 from "@angular/core";
import * as i1 from "./component-loader.service";
export class ModalService {
    /**
     * @param {?} componentLoader
     */
    constructor(componentLoader) {
        this.componentLoader = componentLoader;
    }
    /**
     * @param {?} preview
     * @return {?}
     */
    openWithPreview(preview) {
        this.draw();
        this.modalRef.instance.setPreview(preview);
    }
    /**
     * @return {?}
     */
    draw() {
        this.backdropRef = this.componentLoader.load(BackdropComponent);
        this.modalRef = this.componentLoader.load(ModalComponent);
    }
    /**
     * @return {?}
     */
    remove() {
        this.modalRef.destroy();
        this.backdropRef.destroy();
    }
}
ModalService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
/** @nocollapse */
ModalService.ctorParameters = () => [
    { type: ComponentLoaderService }
];
/** @nocollapse */ ModalService.ngInjectableDef = i0.defineInjectable({ factory: function ModalService_Factory() { return new ModalService(i0.inject(i1.ComponentLoaderService)); }, token: ModalService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    ModalService.prototype.modalRef;
    /**
     * @type {?}
     * @private
     */
    ModalService.prototype.backdropRef;
    /**
     * @type {?}
     * @private
     */
    ModalService.prototype.componentLoader;
}
