// ngx-feedback/esm2015/lib/services/orchestrator.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { ModalService } from './modal.service';
import { SelectorsService } from './selectors.service';
import { EventsService } from './events.service';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "./modal.service";
import * as i2 from "./selectors.service";
import * as i3 from "./events.service";
export class OrchestratorService {
    /**
     * @param {?} modalService
     * @param {?} selectorsService
     * @param {?} eventsService
     */
    constructor(modalService, selectorsService, eventsService) {
        this.modalService = modalService;
        this.selectorsService = selectorsService;
        this.eventsService = eventsService;
        this.subscriptions = [];
        this.onSend = new Subject();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.subscriptions.push(this.onFeedbackButtonClick(), this.onSpecificFeedbackClick(), this.onGenericFeedbackClick(), this.onElementSelected(), this.onCloseClick(), this.onSendClick());
    }
    /**
     * @private
     * @return {?}
     */
    onElementSelected() {
        return this.selectorsService.preview.subscribe((/**
         * @param {?} prev
         * @return {?}
         */
        prev => {
            this.modalService.openWithPreview(prev);
        }));
    }
    /**
     * @private
     * @return {?}
     */
    onFeedbackButtonClick() {
        return this.eventsService.feedbackClickObservable.subscribe((/**
         * @return {?}
         */
        () => {
            this.modalService.draw();
        }));
    }
    /**
     * @private
     * @return {?}
     */
    onSpecificFeedbackClick() {
        return this.eventsService.specificFeedbackClickObservable.subscribe((/**
         * @return {?}
         */
        () => {
            this.modalService.remove();
            this.selectorsService.elementsHighlight();
        }));
    }
    /**
     * @private
     * @return {?}
     */
    onGenericFeedbackClick() {
        return this.eventsService.genericFeedbackClickObservable.subscribe((/**
         * @return {?}
         */
        () => {
            this.modalService.remove();
            this.selectorsService.pageScreenshot();
        }));
    }
    /**
     * @private
     * @return {?}
     */
    onCloseClick() {
        return this.eventsService.closeClickClickObservable.subscribe((/**
         * @return {?}
         */
        () => {
            this.modalService.remove();
        }));
    }
    /**
     * @private
     * @return {?}
     */
    onSendClick() {
        return this.eventsService.onSendClickObservable.subscribe((/**
         * @param {?} data
         * @return {?}
         */
        (data) => {
            this.onSend.next(data);
        }));
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.subscriptions.forEach((/**
         * @param {?} sub
         * @return {?}
         */
        sub => {
            sub.unsubscribe();
        }));
    }
}
OrchestratorService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
/** @nocollapse */
OrchestratorService.ctorParameters = () => [
    { type: ModalService },
    { type: SelectorsService },
    { type: EventsService }
];
/** @nocollapse */ OrchestratorService.ngInjectableDef = i0.defineInjectable({ factory: function OrchestratorService_Factory() { return new OrchestratorService(i0.inject(i1.ModalService), i0.inject(i2.SelectorsService), i0.inject(i3.EventsService)); }, token: OrchestratorService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    OrchestratorService.prototype.subscriptions;
    /** @type {?} */
    OrchestratorService.prototype.onSend;
    /**
     * @type {?}
     * @private
     */
    OrchestratorService.prototype.modalService;
    /**
     * @type {?}
     * @private
     */
    OrchestratorService.prototype.selectorsService;
    /**
     * @type {?}
     * @private
     */
    OrchestratorService.prototype.eventsService;
}
