// ngx-feedback/esm2015/lib/services/selectors.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, RendererFactory2 } from '@angular/core';
import { fromEvent, Subject } from 'rxjs';
import html2canvas from 'html2canvas';
import * as i0 from "@angular/core";
export class SelectorsService {
    /**
     * @param {?} rendererFactory
     */
    constructor(rendererFactory) {
        this.rendererFactory = rendererFactory;
        this.preview = new Subject();
        this.subs = [];
        this.renderer = this.rendererFactory.createRenderer(null, null);
    }
    /**
     * @return {?}
     */
    pageScreenshot() {
        this.screenshot(document.body);
    }
    /**
     * @return {?}
     */
    elementsHighlight() {
        /** @type {?} */
        let element;
        /** @type {?} */
        const elements = [];
        this.source = fromEvent(document, 'mousemove').subscribe((/**
         * @param {?} ev
         * @return {?}
         */
        (ev) => {
            if (element) {
                this.renderer.removeClass(element, 'ngx-feedback__highlight');
            }
            element = document.elementFromPoint(ev.x, ev.y);
            if (elements.findIndex((/**
             * @param {?} el
             * @return {?}
             */
            el => el === element)) === -1) {
                elements.push(element);
                this.subs.push(fromEvent(element, 'click').subscribe((/**
                 * @param {?} evt
                 * @return {?}
                 */
                (evt) => {
                    evt.stopPropagation();
                    this.screenshot(element);
                })));
            }
            this.renderer.addClass(element, 'ngx-feedback__highlight');
        }));
    }
    /**
     * @param {?} element
     * @return {?}
     */
    screenshot(element) {
        /** @type {?} */
        let image;
        html2canvas(element).then((/**
         * @param {?} canvas
         * @return {?}
         */
        canvas => {
            image = ((/** @type {?} */ (canvas))).toDataURL();
            this.preview.next(image);
        }));
        this.renderer.removeClass(element, 'ngx-feedback__highlight');
        this.clean();
    }
    /**
     * @return {?}
     */
    clean() {
        if (this.source) {
            this.source.unsubscribe();
        }
        this.subs.forEach((/**
         * @param {?} sub
         * @return {?}
         */
        sub => sub.unsubscribe()));
    }
}
SelectorsService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
/** @nocollapse */
SelectorsService.ctorParameters = () => [
    { type: RendererFactory2 }
];
/** @nocollapse */ SelectorsService.ngInjectableDef = i0.defineInjectable({ factory: function SelectorsService_Factory() { return new SelectorsService(i0.inject(i0.RendererFactory2)); }, token: SelectorsService, providedIn: "root" });
if (false) {
    /** @type {?} */
    SelectorsService.prototype.preview;
    /** @type {?} */
    SelectorsService.prototype.subs;
    /** @type {?} */
    SelectorsService.prototype.source;
    /**
     * @type {?}
     * @private
     */
    SelectorsService.prototype.renderer;
    /**
     * @type {?}
     * @private
     */
    SelectorsService.prototype.rendererFactory;
}
